hardware/MQTT.cpp

```diff
@@ -945,3 +979,796 @@ void MQTT::SendSceneInfo(const uint64_t SceneIdx, const std::string & /*SceneNam
+				}
+				else
+				{
+					// Update
+
+					if (bRetained)
+						return; // only update when a new value is received
+					UpdateValueInt(m_HwdID, szDeviceID.c_str(), 1, devType, subType, pSensor->SignalLevel, pSensor->BatteryLevel, nValue, sValue.c_str(), result[0][0]);
+				}
+			}
+		}
+	}
+}
+
+void MQTT::InsertUpdateSwitch(_tMQTTASensor *pSensor)
+{
+	pSensor->devUnit = 1;
+	pSensor->devType = pTypeGeneralSwitch;
+	pSensor->subType = sSwitchGeneralSwitch;
+
+	int level = 0;
+	if (level > 100)
+		level = 100;
+
+	int nValue = 0;
+	std::string sValue = std_format("%d", level);
+
+	std::vector<std::vector<std::string>> result;
+	result = m_sql.safe_query("SELECT ID,Name,nValue,sValue FROM DeviceStatus WHERE (HardwareID==%d) AND (DeviceID=='%q')", m_HwdID, pSensor->unique_id.c_str());
+	if (result.empty())
+	{
+		// New switch, add it to the system
+		m_sql.safe_query("INSERT INTO DeviceStatus (HardwareID, DeviceID, Unit, Type, SubType, SignalLevel, BatteryLevel, Name, Used, nValue, sValue) "
+				 "VALUES (%d, '%q', 1, %d, %d, %d, %d, '%q', %d, %d, '%q')",
+				 m_HwdID, pSensor->unique_id.c_str(), pSensor->devType, pSensor->subType, pSensor->SignalLevel, pSensor->BatteryLevel, pSensor->name.c_str(), 1, nValue,
+				 sValue.c_str());
+		result = m_sql.safe_query("SELECT ID,Name,nValue,sValue FROM DeviceStatus WHERE (HardwareID==%d) AND (DeviceID=='%q')", m_HwdID, pSensor->unique_id.c_str());
+	}
+	if (result.empty())
+		return; //should not happen!
+
+	std::string szIdx = result[0][0];
+	uint64_t DevRowIdx = std::stoull(szIdx);
+	std::string szDeviceName = result[0][1];
+	int nvalue = atoi(result[0][2].c_str());
+
+	if (pSensor->last_value.empty())
+		return;
+	bool bOn = false;
+	if (pSensor->last_value == pSensor->payload_on)
+		bOn = true;
+	else if (pSensor->last_value == pSensor->payload_off)
+		bOn = false;
+	else
+		bOn = (pSensor->last_value == "on") || (pSensor->last_value == "ON");
+
+			// check if we have a change, if not do not update it
+	if ((!bOn) && (nvalue == 0))
+		return;
+	if ((bOn && (nvalue != 0)))
+	{
+		// Check Level
+		int slevel = atoi(result[0][3].c_str());
+		if (slevel == level)
+			return;
+	}
+	nValue = (bOn) ? gswitch_sOn : gswitch_sOff;
+	
+	pSensor->nValue = nValue;
+	pSensor->sValue = sValue;
+
+	UpdateValueInt(m_HwdID, pSensor->unique_id.c_str(), pSensor->devUnit, pSensor->devType, pSensor->subType, pSensor->SignalLevel, pSensor->BatteryLevel, pSensor->nValue, pSensor->sValue.c_str(),
+		       szDeviceName);
+
+	// Notify MQTT and various push mechanisms and notifications
+	m_mainworker.sOnDeviceReceived(m_HwdID, DevRowIdx, szDeviceName, nullptr);
+	m_notifications.CheckAndHandleNotification(DevRowIdx, m_HwdID, pSensor->unique_id, szDeviceName, pSensor->devUnit, pSensor->devType, pSensor->subType, pSensor->nValue, pSensor->sValue);
+	m_mainworker.CheckSceneCode(DevRowIdx, pSensor->devType, pSensor->subType, pSensor->nValue, pSensor->sValue.c_str(), "MQTT Auto");
```

Comment:
These notifications are already done in the _UpdateValueInt_ function. No need to repeat them here?

Reply:
Yep thanks for spotting, it was a left over